### PR TITLE
Support for trailing comma

### LINKED ISSUE #28: Trailing comma
Currently Tact doesn't allow tailing comma in struct instantiation:

```
send(SendParameters{
    to: safeAddress,
    value: self.deployValue, // <--- error here
});
```

### PR DESCRIPTION
**Pull request by vitorpy**
Closes #28.
Based on abandoned PR #49.

### CODE REVIEW

File: src/test/features/serialization-2.tact
Comment (anton-trunov) [inline src/test/features/serialization-2.tact:1]: Let's remove this file from this PR completely. You might want to refactor some files in the examples directory, for instance.

**Comment by anton-trunov** [comment]
Hi @vitorpy, thanks for reviving that PR! You made the PR to be a draft, any additional things need to be done here?

**Comment by anton-trunov** [comment]
btw, you probably want to remove src/test/features/serialization-2.tact from your PR -- it's just some unrelated whitespace changes

**Comment by vitorpy** [comment]
> Hi @vitorpy, thanks for reviving that PR! You made the PR to be a draft, any additional things need to be done here?

Hey, I'm going to add support for trailing commas on function arguments lists as well.

File: src/grammar/test/case-25.tact
Comment (anton-trunov) [inline src/grammar/test/case-25.tact:1]: we also need a test for the case when there is zero function parameters: `fun yetAnotherFunction(,)`. This has to be a syntax error

File: src/grammar/grammar.ohm
Comment (anton-trunov) [inline src/grammar/grammar.ohm:1]: It is possible to make the trailing comma optional (`","?`) so we don't effectively duplicate some parts of the grammar? For instance, like so

```
NativeFunction = nameAttribute "(" funcId ")" FunctionAttribute* native id "(" ListOf<FunctionArg,","> ","? ")" ";" --withVoid
```

Comment (anton-trunov) [inline src/grammar/grammar.ohm:1]: We also need to allow trailing comma for `Function` in addition to `NativeFunction`

Comment (anton-trunov) [inline src/grammar/grammar.ohm:1]: Function call syntax `foo(a,b,c)` needs to be adjusted to allow trailing commas too (and have corresponding grammar tests)

**Review by anton-trunov — CHANGES_REQUESTED** [review]
I left a few suggestions that should be pretty easy to fix. Looking forward to having the trailing comma feature in the language!

**Comment by anton-trunov** [comment]
Please add also the corresponding entry in CHANGELOG.md (the `Changed` section)

File: CHANGELOG.md [hunk 1]
```diff
@@ -11,6 +11,7 @@ and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0
 
 ### Changed
 - Update the `dump` function to handle addresses: PR [#175](https://github.com/tact-lang/tact/pull/175)
+- Support for trailing comma: PR [#179](https://github.com/tact-lang/tact/pull/179)
```
Comment (anton-trunov) [inline CHANGELOG.md]:
```suggestion
- Support trailing commas for struct fields and function arguments: PR [#179](https://github.com/tact-lang/tact/pull/179)
```

File: src/grammar/test/case-25.tact
Comment (anton-trunov) [inline src/grammar/test/case-25.tact:1]: We also need a test case for calling a function:
```tact
foo(
  x,
  y,
  z,
)
```

**Review by anton-trunov — CHANGES_REQUESTED** [review]
Awesome stuff! Thanks for all the quick fixes. Just one tiny change and we are good to go

**Comment by anton-trunov** [comment]
I just added this PR to the v1.3.0 milestone so we will make it available in the very next Tact release

**Comment by vitorpy** [comment]
> I just added this PR to the v1.3.0 milestone so we will make it available in the very next Tact release

Thanks! That was fun! I'll see if I grab another issue 😃

File: src/grammar/test/case-25.tact [hunk 2]
```diff
@@ -1,3 +1,18 @@
...
+}
+
+fun anotherFunction(
+    a: Int,
+    b: Int,
+    ): Int {
+    return (a >> b) || (a << (32 - b));
+}
+
+fun oneMoreFunction(
```
Comment (anton-trunov) [inline src/grammar/test/case-25.tact:28]: Ah, sorry, I wasn't perhaps clear enough with the above comment. We don't need more function definitions or declarations, just function calls, something like this:

```tact
fun oneMoreFunction(): Int {
    return foo(                   // <-- function call
        x,
        y,
    );
}
```

**Review by anton-trunov — CHANGES_REQUESTED** [review]

File: src/grammar/grammar.ohm
Comment (anton-trunov) [reply src/grammar/grammar.ohm:1]: This one turns out to be not resolved yet https://github.com/tact-lang/tact/pull/179#discussion_r1532313556

**Comment by anton-trunov** [comment]
> Thanks! That was fun! I'll see if I grab another issue 😃

@vitorpy By all means! If you decide to do it, it would be nice if you outlined your solution before going for it, just to make sure we are on the same page and to make the code review process a faster and smoother experience.

**Comment by vitorpy** [comment]
> > Thanks! That was fun! I'll see if I grab another issue 😃
> 
> @vitorpy By all means! If you decide to do it, it would be nice if you outlined your solution before going for it, just to make sure we are on the same page and to make the code review process a faster and smoother experience.

Will do. This one I went directly as I saw the old PR and it seemed like a good way to get to know the code base a little without too much trouble. Fixed the test case - thanks for the review!

File: src/grammar/grammar.ohm
Comment (anton-trunov) [inline src/grammar/grammar.ohm:1]: I think we should also modify the `ExpressionCall`, `ExpressionStaticCall`, `ExpressionInitOf` non-terminals to allow trailing commas

File: src/grammar/test/case-25.tact
Comment (anton-trunov) [inline src/grammar/test/case-25.tact:1]: In addition to static call we need to make sure `ExpressionCall` and `ExpressionInitOf` are also covered

**Review by anton-trunov — CHANGES_REQUESTED** [review]

File: src/grammar/test/case-25.tact
Comment (vitorpy) [reply src/grammar/test/case-25.tact:1]: I'm not sure how to do this, can you suggest some examples? Thanks!

Comment (anton-trunov) [reply src/grammar/test/case-25.tact:1]: Basically we need to,


1. call an extension method:

```tact
let floatString: String = number.toFloatString(3);
```

2. call `initOf`, for instance:

```tact
fun getNftItemInit(item_index: Int): StateInit {
    return initOf NftItem(collectionAddress, item_index, self.owner_address, self.collection_content);
}
```

File: src/grammar/test-failed/case-19.tact [hunk 3]
```diff
@@ -0,0 +1 @@
+fun function(,) {}
```
Comment (anton-trunov) [inline src/grammar/test-failed/case-19.tact]: this test does not fail

File: src/grammar/test/case-17.tact [hunk 4]
```diff
@@ -1,7 +1,27 @@
...
+    y: Int;
+}
+
+const b: B = new {
+    x: 2,
+    y: 3,
+};
+
+fun getB(): B {
+    return B { x = 1, y = 5, };
```
Comment (anton-trunov) [inline src/grammar/test/case-17.tact]:
```suggestion
    return B { x: 1, y: 5, };
```

**Review by anton-trunov — CHANGES_REQUESTED** [review]
CI shows that some tests are failing

You can see here how to run it locally:
https://github.com/tact-lang/tact/blob/main/.github/workflows/tact.yml#L35-L38

File: src/grammar/test-failed/case-19.tact (hunk 3 above)
Comment (anton-trunov) [inline src/grammar/test-failed/case-19.tact]: We have one test case for the `Function_withVoid` non-terminal, we also need to cover the rest of them:
- `Function_withType`,
- `Function_abstractVoid`,
- `Function_abstractType`,
- `NativeFunction_withType`,
- and `NativeFunction_withVoid`.

File: src/grammar/test-failed/case-19.tact
Comment (anton-trunov) [inline src/grammar/test-failed/case-19.tact:1]: We need test cases not only for function definitions, but also for function *calls*: `... return foo(,);`

Comment (anton-trunov) [inline src/grammar/test-failed/case-19.tact:1]: And a failing test for method calls: `... return 42.toString(,);`

**Review by anton-trunov — CHANGES_REQUESTED** [review]
Yay, it passes the existing tests! It would be great to add more test cases for all the modified grammatical categories to make sure we are consistent and not missing anything

File: src/grammar/test-failed/case-19.tact
Comment (vitorpy) [reply src/grammar/test-failed/case-19.tact:1]: Expanded coverage and checks!

File: src/grammar/__snapshots__/grammar.spec.ts.snap [hunk 5]
```diff
@@ -187,6 +187,35 @@ exports[`grammar should fail case-19 1`] = `
...
+"<unknown>:1:14: Empty parameter list should not have a dangling comma.
+Line 1, col 14:
+> 1 | fun function(,) : Int {
+                   ^
+  2 |     return 42;
+"
+`;
+
+exports[`grammar should fail case-22 1`] = `
+"<unknown>:1:4: Syntax error: expected not an idPart
```
Comment (anton-trunov) [inline src/grammar/__snapshots__/grammar.spec.ts.snap]: > Syntax error: expected not an idPart

This is not a really great error message, can we perhaps do better here?

File: src/grammar/__snapshots__/grammar.spec.ts.snap
Comment (anton-trunov) [inline src/grammar/__snapshots__/grammar.spec.ts.snap:1]: Please add a `Function_abstractVoid` failing test

Comment (anton-trunov) [inline src/grammar/__snapshots__/grammar.spec.ts.snap:1]: Please add a `Function_abstractType` failing test

Comment (anton-trunov) [inline src/grammar/__snapshots__/grammar.spec.ts.snap:1]: Please add a `NativeFunction_withType` failing test

Comment (anton-trunov) [inline src/grammar/__snapshots__/grammar.spec.ts.snap:1]: Please add a `NativeFunction_withVoid ` failing test

**Review by anton-trunov — CHANGES_REQUESTED** [review]

**Comment by vitorpy** [comment]
Sorry for all the back and forth, I think I finally got the hang of writing and debugging tests for tact. Thanks!

File: src/grammar/__snapshots__/grammar.spec.ts.snap [hunk 6]
```diff
@@ -211,11 +211,22 @@ Line 1, col 14:
...
+Line 6, col 14:
+  5 | fun b(): Int {
+> 6 |     return a(,);
+                   ^
+  7 | }
+"
+`;
+
+exports[`grammar should fail case-23 1`] = `
+"<unknown>:7:1: Abstract function doesn't have abstract modifier
```
Comment (anton-trunov) [inline src/grammar/__snapshots__/grammar.spec.ts.snap]: Looks like a different error message than I would have expected

**Comment by anton-trunov** [comment]
> Sorry for all the back and forth, I think I finally got the hang of writing and debugging tests for tact. Thanks!

Hey @vitorpy, no worries at all! Thank you for bringing this feature to Tact and helping to make the implementation more robust

File: src/grammar/__snapshots__/grammar.spec.ts.snap (hunk 6 above)
Comment (vitorpy) [reply src/grammar/__snapshots__/grammar.spec.ts.snap]: Missed that, sorry! Fixed.

File: src/grammar/__snapshots__/grammar.spec.ts.snap [hunk 7]
```diff
@@ -182,6 +182,54 @@ Line 2, col 15:
...
+                   ^
+  7 | }
+"
+`;
+
+exports[`grammar should fail case-23 1`] = `
+"<unknown>:7:31: Empty parameter list should not have a dangling comma.
+Line 7, col 31:
+  6 | 
+> 7 | abstract fun testFuncAbstract(,);
```
Comment (anton-trunov) [inline src/grammar/__snapshots__/grammar.spec.ts.snap]: I only see one error here, but the `test-failed/case-23` contains several of them

Comment (vitorpy) [reply src/grammar/__snapshots__/grammar.spec.ts.snap]: Definitely doing small test files next time to avoid missing that :-)

**Review by anton-trunov — APPROVED** [review]

File: src/grammar/test-failed/case-24.tact
Comment (anton-trunov) [inline src/grammar/test-failed/case-24.tact:1]: let's split this file into two test cases

File: src/grammar/test-failed/case-23.tact
Comment (anton-trunov) [inline src/grammar/test-failed/case-23.tact:1]: let's split this file into two test cases

**Review by anton-trunov — CHANGES_REQUESTED** [review]
CI is red unfortunately

**Comment by vitorpy** [comment]
@anton-trunov I'm kinda stuck on why case-0 is not throwing. Could you give me some pointers on how to progress?

File: src/grammar/grammar.ts [hunk 8]
```diff
@@ -327,35 +339,45 @@ semantics.addOperation<ASTNode>('resolve_declaration', {
...
     },
-    Function_abstractType(arg0, _arg1, arg2, _arg3, arg4, _arg5, _arg6, arg7, _arg8) {
+    Function_abstractType(arg0, _arg1, arg2, _arg3, arg4, arg5, _arg6, _arg7, arg8, _arg9) {
+        if (arg4.source.contents === '' && arg5.sourceString === ',') {
+            throwError('Empty parameter list should not have a dangling comma.', createRef(arg5));
+        }
+
         const attributes = arg0.children.map((v) => v.resolve_attributes()) as ASTFunctionAttribute[];
-        checkVariableName(arg2.sourceString, createRef(arg2));
-        checkFunctionAttributes(true, attributes, createRef(this));
```
Comment (anton-trunov) [inline src/grammar/grammar.ts]: Looks like you inadvertently deleted these lines and it caused that test to fail

File: src/grammar/test-failed/case-26.tact [hunk 9]
```diff
@@ -0,0 +1,3 @@
+
```
Comment (anton-trunov) [inline src/grammar/test-failed/case-26.tact]:
```suggestion
```

File: src/grammar/test-failed/case-25.tact [hunk 10]
```diff
@@ -0,0 +1,2 @@
+
```
Comment (anton-trunov) [inline src/grammar/test-failed/case-25.tact]:
```suggestion
```

**Review by anton-trunov — DISMISSED** [review]

**Review by anton-trunov — DISMISSED** [review]

**Review by anton-trunov — APPROVED** [review]

**Comment by anton-trunov** [comment]
@vitorpy Thank you for your contribution! This qualifies for the Tact contributor SBT: https://society.ton.org/contribute-to-tact-compiler. I'll just need your TON wallet address to issue the token

**Comment by vitorpy** [comment]
Awesome! I'll try to keep the contributions going. 

UQAz6vmh0aKPSH5dJHsfHI8YQH55oLeuWMz4eQkr4ZjS-aG6

**Comment by anton-trunov** [comment]
SBT issued! Hope to see more things from you 👍
